sdk/src/client/mod.rs

```diff
@@ -0,0 +1,243 @@
+                            chunk,
+                        )?
+                        .map_err(|msg| format_err!("server error: {msg}"))?;
+
+                    buf.drain(..index);
+                    if let Some(data) = data {
+                        for item in data {
+                            y.send(Ok(item)).await;
+                        }
+                    } else {
+                        return Ok(());
+                    }
+                }
+            }
+            bail!("unexpected end of response");
+        })
+        .boxed()
+    }
+
+    #[instrument(skip_all, fields(?hash))]
+    pub async fn upload(
+        &self,
+        hash: &EncryptedContentHash,
+        mut encrypted_file: impl Read + Seek + Send + 'static,
+    ) -> Result<()> {
+        let size = encrypted_file.seek(SeekFrom::End(0))?;
+        let encrypted_file = Arc::new(Mutex::new(encrypted_file));
+        let url = self.content_url(hash)?;
+        ok_or_retry(|| async {
+            encrypted_file
+                .lock()
+                .await
+                .rewind()
+                .map_err(RequestError::application)?;
+            self.reqwest
+                .put(url.clone())
+                .timeout(upload_timeout(size))
+                .bearer_auth(self.token.as_ref())
+                .header(CONTENT_LENGTH, size)
+                .body(Body::wrap_stream(
+                    stream_file(encrypted_file.clone()).map(io::Result::Ok),
+                ))
+                .send()
+                .await
+                .map_err(RequestError::transport)?
+                .error_for_status()
+                .map_err(RequestError::application)?;
+            Ok(())
+        })
+        .await
+    }
+
+    fn content_url(&self, hash: &EncryptedContentHash) -> Result<Url> {
+        let mut url = self.server_url.clone();
+        url.path_segments_mut()
+            .map_err(|()| format_err!("failed server URL extension"))?
+            .push("content")
+            .push(&hash.to_url_safe());
+        Ok(url)
+    }
+}
+
+fn take_chunk(buf: &[u8]) -> Option<(&[u8], usize)> {
+    if buf.len() < 4 {
+        return None;
+    }
+    let len = LE::read_u32(buf) as usize;
+    if buf.len() < 4 + len {
+        return None;
+    }
+    Some((&buf[4..4 + len], 4 + len))
+}
+
+/// Retries the request if an error arises due to the transport.
+async fn ok_or_retry<T, F, Fut>(mut f: F) -> Result<T>
+where
+    F: FnMut() -> Fut,
+    Fut: Future<Output = Result<T, RequestError>>,
+{
+    const NUM_RETRIES: usize = 5;
+    const RETRY_PERIOD: Duration = Duration::from_secs(10);
+    let mut attempt = 0;
+    loop {
+        attempt += 1;
+        let err = match f().await {
+            Ok(x) => break Ok(x),
+            Err(RequestError::Application(err)) => break Err(err),
+            Err(RequestError::Transport(err)) if attempt >= NUM_RETRIES => break Err(err),
+            Err(RequestError::Transport(err)) => err,
```

Comment:
I'm not a big fan of condition guards in `match` because it can be hard to reason about exact conditions for branch execution. I'd prefer this to be:
```suggestion
            Err(RequestError::Transport(err)) => {
                if attempt >= NUM_RETRIES {
                    break Err(err);
                } else {
                    err
                }
            },
```